[PEAUTY-210] Add Logic for Process and Thread State Changes on Payment Completion

harvartz opened the pull request:
## 📄 [PF-210] Add Logic for Process and Thread State Changes on Payment Completion

Jira : [PEAUTY-210](https://multicampusuplus.atlassian.net/browse/PEAUTY-210?atlOrigin=eyJpIjoiNmY0YjllNTUyNDExNDdmMTg0ZmE1MTUwYTQ0MGQ1ZTEiLCJwIjoiaiJ9)

## ✨ 변경 사항
- [x] 기능 추가/변경 설명
- [x] 버그 수정 설명
- [x] 문서 수정 설명

<!-- Pull Request의 설명을 추가하세요. -->
- 견적서 수락 시, 기존에는 API를 사용하여 프로세스, 쓰레드 상태 변화를 직접 진행했습니다.
- 지금은 결제기능이 있어서 결제가 유효성 검증을 통해 정상적으로 진행되었을 때, 프로세스와 쓰레드 상태 변화를 진행하게 로직을 추가했습니다.
- payment 모듈의 스웨거 관련 라이브러리를 추가했습니다.
- 추후, 결제에 관련된 문서를 온보딩하겠습니다.

- 포트원 가격과 서버에 저장된 가격이 맞지 않으면 예외처리를 진행하여 쓰레드에 상태에도 변화를 주었지만, 결제 실패로 진행하여 다시 결재할 수 있는 상태로 만들었습니다.

## 📅 작업 일정
<!-- 해당 작업을 수행하는데 예상했던 공수와 실제 소요되었던 공수를 기입해주세요. -->
- Expected MD: 1.3MD
- Actual MD: 1.3MD
### Difference reason (If correct, no need.)
- None.

## ✔️ 확인 사항

- [x] 코드가 잘 작동하는지 확인했나요?
- [x] 새로운 기능에 대한 테스트가 추가되었나요?
- [x] 문서가 업데이트되었나요?

harvartz committed:
docs: 주석 제거, 사용하지 않는 controller 주석 처리

harvartz committed:
style: request, response swagger 어노테이션

harvartz committed:
chore: swagger build.gradle 추가

harvartz committed:
feat: order 도메인의 상태 값 변화
- complete, cancel 상태 변화

harvartz committed:
style: swagger docs 설명 추가, pathvariable 추가
- url 수정

harvartz committed:
refactor: 결제에 따른 프로세스, 쓰레드 상태 변화 로직 추가
- before. api를 사용하여 프로세스, 쓰레드 상태 변화
- after. 완료되면 프로세스, 쓰레드 상태 변화를 할 수 있게 로직 추가
- 그에 필요한 BiddingProcessAdapter, Port 추가(다른 패키지에서 사용하는 파일 복사)

Shanate commented on peauty-customer-api/src/main/java/com/peauty/customer/presentation/controller/customer/CustomerController.java:
```diff
@@ -68,20 +65,4 @@ public GetDesignerBadgesForCustomerResponse getDesignerBadgesForCustomer(@PathVa
         return GetDesignerBadgesForCustomerResponse.from(result);
     }
 
-/*    @PostMapping("/users/{userId}/puppies/{puppyId}/bidding/processes/{processId}/threads/{threadId}/reviews")
-    @Operation(summary = "리뷰 작성", description = "고객이 자신의 강아지를 미용한 디자이너에게 리뷰를 작성하는 API 진입점입니다.")
-    public RegisterReviewResponse registerReview(@PathVariable Long userId,
-                                                 @PathVariable Long puppyId,
-                                                 @PathVariable Long threadId,
-                                                 @PathVariable Long processId,
-                                                 @RequestBody RegisterReviewRequest request){
-        RegisterReviewResult result = reviewService.registerReview(
-                userId,
-                puppyId,
-                threadId,
-                processId,
-                request.toCommand());
-        return RegisterReviewResponse.from(result);
-    }*/
-
```
주석 지운것 좋습니다.

weejinyoung replied:
청소 감사합니당

Shanate reviewed (approved):
수고하셨습니다 👍

weejinyoung commented on peauty-customer-api/src/main/java/com/peauty/customer/presentation/controller/bidding/CustomerBiddingController.java:
```diff
@@ -27,22 +27,24 @@ public SendEstimateProposalResponse initProcessWithSendEstimateProposal(
+//                puppyId,
+//                processId,
+//                threadId
+//        );
+//        return AcceptEstimateResponse.from(result);
+//    }
```
감사합니당

weejinyoung commented on peauty-domain/src/main/java/com/peauty/domain/payment/Order.java:
```diff
@@ -34,4 +34,11 @@ public void updateOrderToReady() {
         this.orderStatus = OrderStatus.CANCELLED;
     }
+
+    public void updatePayment(Payment paymentToValidate) {
+        this.payment = paymentToValidate;
+    }
```
이거 어떤 메소드인지 감이 잘 오지 않아욤
update 라는 추상적인 네이밍보단 더 자세한 네이밍 혹시 있을까요?

harvartz replied:
해당 메서드는 유효성 검증이 끝나고 나서 진행할 메서드라 updateValidatedPayment()로 수정했습니다!

weejinyoung commented on peauty-payment/src/main/java/com/peauty/payment/business/PaymentServiceImpl.java:
```diff
@@ -73,30 +82,42 @@ public CompletePaymentCallbackResult completePayment(
             }
 
             // 2. 포트원의 정보와 내가 저장한 값이 같은지 확인하기
-            log.info("payment price: {}\n actualAmount: {}", paymentToValidate.getDepositPrice(), actualAmount);
-            // TODO. chekcIfPaymentPriceEquals 조금 더 직관적인 메서드명으로 리팩토링하기
+            // TODO. checkIfPaymentPriceEquals 조금 더 직관적인 메서드명으로 리팩토링하기
             if (!paymentToValidate.checkIfPaymentPriceEquals(actualAmount)) {
-                order.updateOrderToReady();
+                order.updateOrderToCanceled();
                 paymentToValidate.updateStatusToCancel();
                 paymentPort.savePayment(paymentToValidate);
-
+                processToChangeStatus.cancelThread(new BiddingThread.ID(threadId));
                 iamportClient.cancelPaymentByImpUid(new CancelData(iamportResponse.getImpUid()
```
스레드의 취소란 고객과 디자이너 간의 입찰 관계가 끝나버리는 것인데 결제 검증에 실패했다고 입찰 관계가 끝나진 않을 것입니다
결제 검증에 실패하면 예외를 던져서 트랜잭션을 롤백하는 정도가 맞을 것 같아요!

weejinyoung reviewed (changes requested):
고생하셨습니당 
중요한 로직 관련 리뷰가 포함됐어요 확인 부탁드려요!

harvartz committed:
Merge branch 'release/v0.1.0' into feature/PEAUTY-210

harvartz committed:
refactor: 리뷰 반영

harvartz committed:
Merge branch 'release/v0.1.0' into feature/PEAUTY-210

harvartz committed:
Merge branch 'release/v0.1.0' into feature/PEAUTY-210

harvartz committed:
fix: 기존 예외처리 수정
- 포트원 가격과 서버에 저장된 가격이 맞지 않으면 예외처리를 진행하여 쓰레드에 상태에도 변화를 주었지만, 결제 실패로 진행하여 다시 결재할 수 있는 상태로 만들었음

harvartz committed:
docs: 주석 제거

Shanate reviewed (approved):
LGTM

weejinyoung commented on peauty-payment/src/main/java/com/peauty/payment/business/PaymentServiceImpl.java:
```diff
@@ -56,72 +58,77 @@ public CreateOrderResult saveOrder(
-                paymentPort.savePayment(paymentToValidate);
-                processToChangeStatus.cancelThread(new BiddingThread.ID(threadId));
+            if (paymentToValidate.checkPaymentPriceEqualsWithPortonePrice(actualAmount)) {
+                orderToValidate.updateStatusToError();
+                paymentToValidate.updateStatusToError();
                 iamportClient.cancelPaymentByImpUid(new CancelData(iamportResponse.getImpUid()
                         , true, new BigDecimal(actualAmount)));
-                throw new PeautyException(PeautyResponseCode.PAYMENT_AMOUNT_NOT_EQUALS);
             }
```
예외가 없어졌네요! 감사합니다

weejinyoung reviewed (approved):
어려운 도메인이라 명확하게 이해는 안되지만... 석진 님의 결제 로직 세미나를 다시 한 번 기대하겠습니다 고생 많으셨습니다

harvartz committed:
Merge branch 'release/v0.1.0' into feature/PEAUTY-210
